### PR TITLE
JC figure s12

### PR DESCRIPTION
**Pull request by jclagarde**

### CODE REVIEW

File: lineage/figures/figureS12.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+from ..LineageTree import LineageTree
+
+
+def makeFigure():
+    """
+    Makes figure S12.
+    """
+
+    # Get list of axis objects
+    ax, f = getSetup((10, 10), (1, 3))
```
Comment (Farnazmdi) [inline lineage/figures/figureS12.py:34]: (10, 10) gives you a square. This is a 1 by 3 figure, so the width should be ~3 times the length.

File: lineage/figures/figureS12.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    tree_obj, predicted_states, LL = Analyze(X_3, 3)
+    results_dict = Results(tree_obj, predicted_states, LL)
+    accuracy = results_dict["accuracy_after_switching"]
+        
+    X4.append(tmp_lineage4)
+    tree_obj, predicted_states, LL = Analyze(X_4, 4)
+    results_dict = Results(tree_obj, predicted_states, LL)
+    accuracy = results_dict["accuracy_after_switching"]
+
+    return accuracy
```
Comment (Farnazmdi) [inline lineage/figures/figureS12.py:106]: The accuracy should be named differently for 2, 3, and 4 states.

Comment (Farnazmdi) [inline lineage/figures/figureS12.py:106]: Here is what `figureMaker` that you imported takes in: ``figureMaker(ax, x, paramEst, accuracies, tr, pii, paramTrues, xlabel="Number of Cells")

When you want to call the function, if you say `figureMaker(*accuracy())` it means all the stuff that `accuracy()` is returning goes to the argument. But what you are returning in your accuracy function only has accuracy.

File: lineage/figures/figureS12.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    a uncensored two-state model.
+    We increase the desired number of cells in a lineage by
+    the experiment time.
+    """
+
+    pi2 = np.array([0.60, 0.40])
+    T2 = np.array([[0.75, 0.25], [0.1, 0.90]])
+    E2 = [StateDistPhase(0.99, 0.9, 20, 5, 10, 3), StateDistPhase(0.88, 0.75, 10, 2, 15, 4)]
+    
+    X_2 = [LineageTree.init_from_parameters(pi2, T2, E2, desired_num_cells=(2 ** 11) - 1)]
```
Comment (Farnazmdi) [inline lineage/figures/figureS12.py:56]: This line, and similarly, X_3 and X_4 that you have here, will create only one lineage. Ideally, we would want to analyze a population with more than 1 lineage. You can look at these lines in figure4.py and do a slight change to create a population with increasing cell number *and* lineage number.
`num_lineages = np.linspace(min_num_lineages, int(0.35 * max_num_lineages), num_data_points, dtype=int)
    experiment_times = np.linspace(1000, int(2.5 * 1000), num_data_points)
    list_of_populations = []
    list_of_fpi = []
    list_of_fT = []
    list_of_fE = []
    for indx, num in enumerate(num_lineages):
        population = []
        for _ in range(num):
            good2go = False
            while not good2go:
                tmp_lineage = LineageTree.init_from_parameters(pi, T, E2, desired_num_cells=min_desired_num_cells, censor_condition=3, desired_experiment_time=experiment_times[indx])
                good2go = lineage_good_to_analyze(tmp_lineage)
            population.append(tmp_lineage)
        # Adding populations into a holder for analysing
        list_of_populations.append(population)`

**Comment by jclagarde** [comment]
Screen Shot 2020-08-17 at 11 39 44 PM
I keep receiving the error `'LineageTree' object is not subscriptable` I've been trying to figure out why and compare my code to figure4.py and trying different functions and I'm still not sure what's causing it

**Comment by Farnazmdi** [comment]
The `X` that you pass to the `Analyze` should be a list where you have appended lineages. If you have one lineage, you should pass it to the `Analyze` in a list like this: `[lineage_tree]`.
@jclagarde

**Comment by Farnazmdi** [comment]
Okay, so follow my changes. I made changes only for  2-state chunk, you can copy and paste that for 3 and 4-state.
You will have `cell_number_x2` for two states, one of the outputs of `commonAnalyze()`. You will need the accuracy and the cell_number for 2, 3, and 4 states that are two of the outputs of `commonAnalyze`.

Then follow the structure to create a dataframe out of them. Then, complete the figureMaker function for the other two.
We can't use the `figureMaker` from `figureCommons` because our figure's structure is different from that of `figureCommons`.

Let me know if anything doesn't make sense. (The run fails now because I intentionally left couple of things incomplete for you.)
@jclagarde

**Comment by Farnazmdi** [comment]
For the two-state, you can actually use the imported `pi`, `T` and `E`, and get rid of `pi2`, `T2`, and `E2`.

**Comment by jclagarde** [comment]
> The `X` that you pass to the `Analyze` should be a list where you have appended lineages. If you have one lineage, you should pass it to the `Analyze` in a list like this: `[lineage_tree]`.
> @jclagarde


> Okay, so follow my changes. I made changes only for 2-state chunk, you can copy and paste that for 3 and 4-state.
> You will have `cell_number_x2` for two states, one of the outputs of `commonAnalyze()`. You will need the accuracy and the cell_number for 2, 3, and 4 states that are two of the outputs of `commonAnalyze`.
> 
> Then follow the structure to create a dataframe out of them. Then, complete the figureMaker function for the other two.
> We can't use the `figureMaker` from `figureCommons` because our figure's structure is different from that of `figureCommons`.
> 
> Let me know if anything doesn't make sense. (The run fails now because I intentionally left couple of things incomplete for you.)
> @jclagarde

@Farnazmdi just saw this! Working on it right now, thank you

File: lineage/figures/figureS12.py [hunk 4]
```diff
@@ -89,16 +85,18 @@ def accuracy():
...
             
             population3.append(tmp_lineage3)
     
-    tree_obj, predicted_states, LL = run_Analyze_over(population3, 3)
-    results_dict = Results(tree_obj, predicted_states, LL)
-    accuracy_3 = results_dict["accuracy_after_switching"]
+    # Adding populations into a holder for analysing
+        list_of_populations2.append(population2)
+
+    cell_number_x2, _, accuracy2_after_switching, _, _, _ = commonAnalyze(list_of_populations2)
```
Comment (Farnazmdi) [inline lineage/figures/figureS12.py:99]: You should change the names of `cell_number_x2` to `cell_number_x3` for 3 state and for 4 state x4, as well.

**Comment by Farnazmdi** [comment]
Did you make the figure locally?

**Comment by jclagarde** [comment]
> Did you make the figure locally?

@Farnazmdi I was trying but it kept giving me the error `operands could not be broadcast together with shapes (2,2) (3,3) ` I realized that it was because commonAnalyze() assumes the lineage is 2 states so I've been trying to find a way to work around it

File: lineage/figures/figureS12.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    
+
+    return accuracy_df
+
+def figureMaker(accuracy_df):
+    """ This creates figure S12. Includes 3 subplots for accuracy of state assignment for 2, 3, and 4 states. """
+
+    i = 0
+    sns.regplot(x="x2", y="accuracy2", data=accuracy_df, ax=ax[i], lowess=True, marker='+', scatter_kws=scatter_kws_list[0])
+    ax[i].set_title("State Assignment Accuracy")
```
Comment (Farnazmdi) [inline lineage/figures/figureS12.py:137]: The title can say how many states we have for each figure.

File: lineage/figures/figureS12.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
+"""
+File: figureS12.py
+Purpose: Generates figure S12.
+Figure S12 shows the accuracy of state assignment versus 
+increasing cell numbers for 2-state, 3-state, and 4-state models.
+"""
+import numpy as np
+import pandas as pd
+from ..Analyze import Analyze, Results, run_Analyze_over
```
Comment (Farnazmdi) [inline lineage/figures/figureS12.py:9]: You can delete this line as you are not using Analyze, results, or run_analyze_over.

**Comment by Farnazmdi** [comment]
figure 3,4,5 are using `commonAnalyze` and you should add the num_state to their argument.

**Comment by Farnazmdi** [comment]
One suggestion. When you are making changes, wait until you think you are done at least with the changes you know you have to make and then *push*. You can commit them one by one, but better to push altogether, this way Jenkins wouldn't have a ton of builds in the queue.

**Comment by jclagarde** [comment]
> One suggestion. When you are making changes, wait until you think you are done at least with the changes you know you have to make and then _push_. You can commit them one by one, but better to push altogether, this way Jenkins wouldn't have a ton of builds in the queue.

ohhh, okay! Thank you for letting me know!

**Comment by jclagarde** [comment]
Screen Shot 2020-08-18 at 2 39 48 PM
@Farnazmdi When you get the chance, can you let me know what you think might be causing this error please?

File: lineage/figures/figureS12.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+def makeFigure():
+    """
+    Makes figure S12.
+    """
+
+    # Get list of axis objects
+    ax, f = getSetup((9, 3), (1, 3))
+
+    figureMaker(ax, *accuracy())
+
```
Comment (Farnazmdi) [inline lineage/figures/figureS12.py:36]: replace this and see if it works:
acc_df = accuracy()
figureMaker(ax, acc_df)

**Comment by Farnazmdi** [comment]
And please add the num_states in `commonAnalyze` that has been called in supplementary figures from S05 to the end.
It is in the very last line.

**Comment by Farnazmdi** [comment]
Do you still have the assertion errors regarding 14 iterations?

**Comment by jclagarde** [comment]
> Do you still have the assertion errors regarding 14 iterations?

@Farnazmdi I do when I change this line under the 4-state population `for _ in range(10):` to `for _ in range(num):`

**Comment by Farnazmdi** [comment]
Hmm.. okay, it probably fails for num=1. Try this line:
`for _ in range(5, num):`
let me know if it changes anything.

**Comment by jclagarde** [comment]
> Hmm.. okay, it probably fails for num=1. Try this line:
> `for _ in range(5, num):`
> let me know if it changes anything.

Screen Shot 2020-08-18 at 4 03 37 PM
Screen Shot 2020-08-18 at 4 04 33 PM
yeah, it's still giving me these errors for some reason

**Comment by Farnazmdi** [comment]
Okay,
I will make some changes now if you are not in the middle of changing any files.

**Comment by Farnazmdi** [comment]
Please do git pull and greate the figure again.

**Comment by jclagarde** [comment]
Screen Shot 2020-08-18 at 4 32 51 PM
@Farnazmdi it worked! :D

**Comment by Farnazmdi** [comment]
Something is off in the third figure, the cell number should roughly be the same.
Make sure the loop is using `num`. And change `the desired_num_cells` from `min_desired_num_cells` to `0.5*max_deired_num_cells`. And removed importing the min_... to max_...

**Comment by Farnazmdi** [comment]
Make sure you added the new argument to the `commonAnalyze` in all supplementary figures that call it.

**Comment by jclagarde** [comment]
> Something is off in the third figure, the cell number should roughly be the same.
> Make sure the loop is using `num`. And change `the desired_num_cells` from `min_desired_num_cells` to `0.5*max_deired_num_cells`. And removed importing the min_... and import max_...

should it be changed for each state model or only the 4 state model?

**Comment by Farnazmdi** [comment]
All of them. Let me handle that. I will make changes right now.

**Comment by Farnazmdi** [comment]
Okay, it should be ready. Thanks JC!

**Comment by jclagarde** [comment]
no problem! thank you for your help!

**Comment by jclagarde** [comment]
Was it able to build for you locally? I get the error `name 'min_num_lineages' is not defined` I can go ahead and add it if it should be there

**Comment by Farnazmdi** [comment]
Do git pull. I did define it.

**Pull request merged**